### PR TITLE
Ns/integration to sharepoint

### LINKED ISSUE #5: Integration of the app to Sharepoint
- implement CRUD operations between the app and Sharepoint

### PR DESCRIPTION
**Pull request by nicolesigei**
Closes #5 
**Features proposed in this pull request**
Integrating the app to SharePoint to retrieve properties and relationship for a site resource and to return metadata for an item in a list. Using CRUD functions in a repository class that are injected in the controllers.

### CODE REVIEW

File: AssetTracking/AssetTracking/Controllers/OfficeBooksController.cs [hunk 1]
```diff
@@ -1,18 +1,66 @@
...
+        private readonly IGraphSdkHelper _graphSdkHelper;
+        public IOfficeBookRepository _officeBookRepository;
+
+        public OfficeBooksController(IGraphSdkHelper graphSdkHelper)
+        {
+            _graphSdkHelper = graphSdkHelper;
+        }
+        public GraphServiceClient GraphClient { get; private set; }
+
         [Route("")]
```
Comment (titusgicheru) [inline AssetTracking/AssetTracking/Controllers/OfficeBooksController.cs]: Remove this attribute if not in use.

File: AssetTracking/AssetTracking/Controllers/OfficeBooksController.cs [hunk 2]
```diff
@@ -1,18 +1,66 @@
...
         public IActionResult OfficeBooks()
         {
+            if (User.Identity.IsAuthenticated)
+            {
+                GraphClient = _graphSdkHelper.GetAuthenticatedClient((ClaimsIdentity)User.Identity);
+                var user = GraphClient.Me.Request().GetAsync();
+            }
             return View("~/Views/Admin/OfficeBooks.cshtml");
```
Comment (titusgicheru) [inline AssetTracking/AssetTracking/Controllers/OfficeBooksController.cs]: Is there a reason why we are explicitly redirecting to the same view? And is this the best approach?

File: AssetTracking/AssetTracking/Controllers/OfficeBooksController.cs [hunk 3]
```diff
@@ -1,18 +1,66 @@
...
         }
+
+        [Route("/GetBooks")]
+        [HttpPost]
+        public async Task<IActionResult> GetBooks()
+        {
+            List<OfficeBook> officeBook = await _officeBookRepository.GetBooks();
+            ViewBag.List = officeBook;
```
Comment (titusgicheru) [inline AssetTracking/AssetTracking/Controllers/OfficeBooksController.cs]: Consider returning a List of **OfficeBook** as opposed to using a **ViewBag**

File: AssetTracking/AssetTracking/Controllers/OfficeBooksController.cs [hunk 4]
```diff
@@ -1,18 +1,66 @@
...
+            return View();
+        }
+
+        [Route("/AddBooks")]
+        [HttpPost]
+        [AutoValidateAntiforgeryToken]
+        public async Task<bool> AddBook(OfficeBook officeBook)
+        {
+            bool result = await _officeBookRepository.AddBook(officeBook);
+            return result;
```
Comment (titusgicheru) [inline AssetTracking/AssetTracking/Controllers/OfficeBooksController.cs]: Perhaps redirect to **GetBooks()** which returns a list of all books.

File: AssetTracking/AssetTracking/Controllers/OfficeBooksController.cs [hunk 5]
```diff
@@ -1,18 +1,66 @@
...
+        }
+
+        [HttpPut]
+        public async Task<bool> UpdateBook(OfficeBook officeBook)
+        {
+            bool result = await _officeBookRepository.UpdateBook(officeBook);
+            return result;
```
Comment (titusgicheru) [inline AssetTracking/AssetTracking/Controllers/OfficeBooksController.cs]: Redirect to a List of all books or to a details View.

File: AssetTracking/AssetTracking/Controllers/OfficeItemsController.cs [hunk 6]
```diff
@@ -1,18 +1,65 @@
...
+                graphserviceClient = _graphSdkHelper.GetAuthenticatedClient((ClaimsIdentity)User.Identity);
+                var user = graphserviceClient.Me.Request().GetAsync();
+            }
+            return View();
+        }
+       
+        public async Task<IActionResult> GetItems()
+        {
+            List<OfficeItem> officeItem = await _officeItemRepository.GetItems();
+            ViewBag.List = _officeItemRepository;
```
Comment (titusgicheru) [inline AssetTracking/AssetTracking/Controllers/OfficeItemsController.cs]: Here also consider returning a List of **OfficeItem** as opposed to using a **ViewBag**

File: AssetTracking/AssetTracking/Models/OfficeBookRepository.cs [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+using Newtonsoft.Json;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace AssetTracking.Models
+{
+    public class OfficeBookRepository : IOfficeBookRepository
+    {
+        private readonly string siteId = "m365b267815.sharepoint.com,6e1261a1-6d03-432a-95c0-e1c7705aef5f,f43d258c-ece0-476a-a1c0-018d359817d5";
```
Comment (titusgicheru) [inline AssetTracking/AssetTracking/Models/OfficeBookRepository.cs]: This should not be hard coded have the **siteId** in a settings file.

File: AssetTracking/AssetTracking/Models/OfficeBookRepository.cs [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+        private readonly string itemId = null;
+        public GraphServiceClient GraphClient { get; private set; }
+        public async Task<List<OfficeBook>> GetBooks()
+        {
+            sharePointLists = await Sites.GetLists(GraphClient, siteId);
+            List<OfficeBook> _officeBooksDirectoryList = new List<OfficeBook>();
+
+            if (sharePointLists != null)
+            {
+                var _officeBookList = sharePointLists.Where(x => x.DisplayName.Contains("Office Books")).FirstOrDefault();
```
Comment (titusgicheru) [inline AssetTracking/AssetTracking/Models/OfficeBookRepository.cs]: Avoid using **inference**, replace all **var** keywords with actual return **types**. This applies to all references in this file.

File: AssetTracking/AssetTracking/Models/OfficeItem.cs [hunk 9]
```diff
@@ -0,0 +1,24 @@
+﻿using Newtonsoft.Json;
+using System;
+
+namespace AssetTracking.Models
+{
+    public class OfficeItem
+    {
+
```
Comment (titusgicheru) [inline AssetTracking/AssetTracking/Models/OfficeItem.cs]: Remove white space

File: AssetTracking/AssetTracking/Models/OfficeItemRepository.cs [hunk 10]
```diff
@@ -0,0 +1,118 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.Graph;
+using Newtonsoft.Json;
+
+namespace AssetTracking.Models
+{
+
```
Comment (titusgicheru) [inline AssetTracking/AssetTracking/Models/OfficeItemRepository.cs]: Remove white space.

File: AssetTracking/AssetTracking/Models/OfficeItemRepository.cs [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+                                                      _listId, itemId);
+                return deletebook;
+            }
+            else
+            {
+                return false;
+            }
+        }
+    }
+
```
Comment (titusgicheru) [inline AssetTracking/AssetTracking/Models/OfficeItemRepository.cs]: Remove white space.

File: AssetTracking/AssetTracking/Models/OfficeItemRepository.cs [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+            if (officeItemsLists != null)
+            {
+                var addItem = officeItemsLists.Where(b => b.DisplayName.Contains("Office Items")).FirstOrDefault();
+                string _listId = addItem.Id;
+
+                string itemId = userItemId;
+
+                IDictionary<string, object> data = new Dictionary<string, object>
+                {
+
```
Comment (titusgicheru) [inline AssetTracking/AssetTracking/Models/OfficeItemRepository.cs]: Remove white space.

File: AssetTracking/AssetTracking/Models/OfficeItemRepository.cs [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.Graph;
+using Newtonsoft.Json;
+
+namespace AssetTracking.Models
+{
+
+    public class OfficeItemRepository : IOfficeItemRepository
+    {
+        private readonly string siteId = "m365b267815.sharepoint.com,6e1261a1-6d03-432a-95c0-e1c7705aef5f,f43d258c-ece0-476a-a1c0-018d359817d5";
```
Comment (titusgicheru) [inline AssetTracking/AssetTracking/Models/OfficeItemRepository.cs]: Avoid hard coding values. Store in config file instead.

**Review by titusgicheru — CHANGES_REQUESTED** [review]

File: AssetTracking/AssetTracking/Models/OfficeBookRepository.cs [hunk 14]
```diff
@@ -0,0 +1,121 @@
...
+using System.Threading.Tasks;
+
+namespace AssetTracking.Models
+{
+    public class OfficeBookRepository : IOfficeBookRepository
+    {
+        private readonly string siteId = "m365b267815.sharepoint.com,6e1261a1-6d03-432a-95c0-e1c7705aef5f,f43d258c-ece0-476a-a1c0-018d359817d5";
+        private ISiteListsCollectionPage sharePointLists;
+        private readonly string itemId = null;
+        public GraphServiceClient GraphClient { get; private set; }
```
Comment (irvinesunday) [inline AssetTracking/AssetTracking/Models/OfficeBookRepository.cs]: Do we really need this property?

File: AssetTracking/AssetTracking/Models/OfficeBookRepository.cs [hunk 15]
```diff
@@ -0,0 +1,121 @@
...
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace AssetTracking.Models
+{
+    public class OfficeBookRepository : IOfficeBookRepository
+    {
+        private readonly string siteId = "m365b267815.sharepoint.com,6e1261a1-6d03-432a-95c0-e1c7705aef5f,f43d258c-ece0-476a-a1c0-018d359817d5";
+        private ISiteListsCollectionPage sharePointLists;
+        private readonly string itemId = null;
```
Comment (irvinesunday) [inline AssetTracking/AssetTracking/Models/OfficeBookRepository.cs]: This field is not used; you can remove it.

File: AssetTracking/AssetTracking/Models/OfficeBookRepository.cs (hunk 8 above)
Comment (irvinesunday) [reply AssetTracking/AssetTracking/Models/OfficeBookRepository.cs]: Additionally, let's maintain uniform variable naming conventions throughout. Preferably an '_' prefixing a field name (followed by camelCasing) for private fields and then camelCase for all method variables.

File: AssetTracking/AssetTracking/Models/OfficeBookRepository.cs [hunk 16]
```diff
@@ -0,0 +1,121 @@
...
+            }
+            return _officeBooksDirectoryList;
+        }
+        public async Task<bool> AddBook(OfficeBook officeBook, GraphServiceClient graphClient)
+        {
+            sharePointLists = await Sites.GetLists(graphClient, siteId);
+
+            if (sharePointLists != null)
+            {
+                var _officeBookList = sharePointLists.Where(b => b.DisplayName.Contains("Office Books")).FirstOrDefault();
```
Comment (irvinesunday) [inline AssetTracking/AssetTracking/Models/OfficeBookRepository.cs]: You can have the "Office Books" defined as a const field data, e.g.
`private const string OfficeBooksDisplayName = "Office Books";`
And then use this across the file where you are checking for the book's display name property.

**Review by irvinesunday — CHANGES_REQUESTED** [review]

**Pull request merged**